CONTRIBUTING.md

```diff
@@ -0,0 +1,232 @@
+2. Use the pull request for design feedback and for iterating on the design.
+3. Once the design is approved, create a new issue with a description that includes the final design document.  Include a link to the pull request that was used for discussion.
+4. Close (without merging!) the pull request used for the design discussion.
+
+## Platform Specific Code
+
+- *Prefer cross-platform code to platform-specific code*
+
+  Cross-platform code is more easily reused.  Reusing code reduces the amount of code that must be written, tested, and maintained.
+
+- *Platform specific code, and only platform specific code, should go in `GVFSPlatform`*
+
+  When platform specific code is required, it should be placed in `GVFSPlatform` or one of the platforms it contains (e.g. `IKernelDriver`)
+
+## Tracing and Logging
+
+- *The "Error" logging level is reserved for non-retryable errors that result in I/O failures or the VFS4G process shutting down*
+
+  The expectation from our customers is that when VFS4G logs an "Error" level message in its log file either:
+    * VFS4G had to shut down unexpectedly
+    * VFS4G encountered an issue severe enough that user-initiated I/O would fail.
+
+- *Log full exception stacks*
+
+  Full exception stacks (i.e. `Exception.ToString`) provide more details than the exception message alone (`Exception.Message`). Full exception stacks make root-causing issues easier.
+  
+- *Do not display full exception stacks to users*
+
+  Exception call stacks are not usually actionable for the user.  Users frequently (sometimes incorrectly) assume that VFS4G has crashed when shown a full stack.  The full stack *should* be included in VFS4G logs, but *should not* be displayed as part of the error message provided to the user.
+
+- *Include relevant details when logging exceptions*
+
+  Sometimes an exception call stack alone is not enough to root cause failures in VFS4G.  When catching (or throwing) exceptions, log relevant details that will help diagnose the issue.  As a general rule, the closer an exception is caught to where it's thrown, the more relevant details there will be to log.
+
+  Example:
+  ```
+  catch (Exception e)
+  {
+    EventMetadata metadata = new EventMetadata();
+    metadata.Add("Area", "Mount");
+    metadata.Add(nameof(enlistmentHookPath), enlistmentHookPath);
+    metadata.Add(nameof(installedHookPath), installedHookPath);
+    metadata.Add("Exception", e.ToString());
+    context.Tracer.RelatedError(metadata, $"Failed to compare {hook.Name} version");
+  ```
+
+## Error Handling
+
+- *Fail fast: An error or exception that risks data loss or corruption should shut down VFS4G immediately*
+
+  Preventing data loss and repository corruption is critical.  If an error or exception occurs that could lead to data loss it's better to shut down VFS4G than keep the repository mounted and risk corruption.
```

Comment:
I think this flows better with a comma?

If an error or exception occurs that could lead to data loss **,** it's better to shut down VFS4G than keep the repository mounted and risk corruption.